File: Nim/Sources/Day03.nim
import strformat
import sequtils
import parseutils
import AoCLib

const InputData = "Day03.txt"

type
  State =tuple[ Data: seq[seq[char]], Transposed :seq[seq[char]]]

let s = new State
s.Data = toSeq((RawDataPath2021 & InputData).lines).mapit(it.toseq())
s.Transposed = s.Data.Transpose()

proc Part01() =
    var
        myOnes:int
        myGamma,myEpsilon:string

    for myitem in s.Transposed:
        myOnes=myitem.filterIt(it=='1').len()
        # This part works when the bias is to '1' if '1' and '0' are equal
        # so we can't use >=, just >
        if myOnes>myitem.len()-myOnes:
            myGamma &= "1"
            myEpsilon &= "0"
        else:
            myGamma &= "0"
            myEpsilon &= "1"
    
    var mygammaint, myepsilonint:int
    discard parseBin(myGamma,mygammaint) 
    discard parsebin(myEpsilon,myepsilonint)
    let myResult:int=mygammaint * myepsilonint
    echo fmt"The answer to Day 03 part 1 is 845186 .  Found is {myResult}"

#In part 2, the bit criteria is calculated and then the readings seq is filered and the transposed array
# regenerated with the filtered dataset
proc Part02() =
    var 
        myOnes:int
        myLastOxygen, myLastCO2:seq[char]
        myReadings:seq[seq[char]]
        myBit:char

    myreadings =s.Data[0..^1]
    s.Transposed = s.Data.Transpose()

    for myIndex in 0..s.Transposed.high:
        mylastoxygen = myReadings[^1]
        myones = s.Transposed[myIndex].countIt(it == '1')
        
        if myones >= myReadings.len() - myones:
            mybit='1'
        else:
            myBit='0'

        myReadings=myReadings.filterit(it[myindex] == myBit)[0..^1]
        #echo fmt"{myOnes},{mybit}, {myreadings.len()}"
        if myReadings.len()==1:
            mylastoxygen = myReadings[0]
            break
        s.Transposed=myReadings.Transpose


    myReadings=s.Data[0..^1]
    s.Transposed = s.Data.Transpose()

    for myIndex in 0..s.Transposed.high:
        mylastco2 = myReadings[^1]
        myones = s.Transposed[myIndex].countIt(it == '1')
        if  myOnes >= myreadings.len() - myones:
            mybit='0'
        else:
            myBit='1'
        
        myReadings= myReadings.filterit(it[myindex]==myBit)[0..^1]
       
        if myReadings.len()==1:
            mylastco2 = myReadings[0]
            break
        s.Transposed=myReadings.Transpose

    var mylastoxygenint, mylastco2int:int
    discard parseBin($mylastoxygen,mylastoxygenint) 
    discard parsebin($mylastco2,mylastco2int)
    let myResult:int=mylastoxygenint * mylastco2int

    #var myResult:int=BinStrToInt(myLastOxygen) * BinStrToInt(myLastCO2)
    echo fmt"The answer to Day 03 Part 2 is 4636702. Found {myResult}"

proc Execute*()=
    
    Part01()
    Part02()

File: Cshp/AoC2021/Day03.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AoC2021CSharp;

namespace AoC2021CSharp
{
    internal static class Day03
    {
        const string InputData= "Day02.txt";

        struct State
        {
            //List<long> integers = new List<long>();
            private List<List<char>> _Data;
            public List<List<char>> Data 
            { 
                get {return _Data;}
                
                set 
                {
                    _Data = value;
                    Transpose();
                } 
            }
            public List<List<char>> Transposed {get;}

            private List<List<char>> Transpose()
            {
                List<List<char>> myTransposed =new List<List<char>>();
               
                foreach( char myChar in Data[0])
                {                    
                    myTransposed.Add(new List<char>());
                        
                }
                    
                foreach (List<char> myList in Data)
                {        
                    for(int myIndex = 0;myIndex <myList.Count; myIndex++)
                    {        
                        myTransposed[myIndex].Add(myList[myIndex]);                         
                    }
                }
                    
                return myTransposed;
            }
        }

        static State s = default;
       


        public static void Execute()
        {

            Part01();
            Part02();

        }

        static void Part01()
        {
            s.Data = new List<List<char>>(File.ReadAllLines(Constants.AoC2021Data + InputData).Select(x => x.ToList<char>()).ToList());

            long myOnes=default;
            string myGamma=default;
            string myEpsilon =default;

            
            foreach (List<char> myItem in s.Transposed)
            {
                myOnes = myItem.Count(x=> x=='1');
                // This part works when the bias is to '1' if ( '1' and '0' are equal
                //so we can't use >=, just >
                if ( myOnes > myItem.Count - myOnes)
                {
                    myGamma = myGamma + '1';
                }
                else
                {
                    myGamma = myGamma + '0';
                    myEpsilon = myEpsilon + '1';
                    
                } 
                
            }
            
            
            long myResult = Convert.ToInt32(myGamma,2) * Convert.ToInt32(myEpsilon,2);

            Console.WriteLine( "The answer to Day 03 part 1 is 845186.  Found is {0}", myResult);
        }

        static void Part02()
        {
       
            s.Data = new List<List<char>>(File.ReadAllLines(Constants.AoC2021Data + InputData).Select(x => x.ToList<char>()).ToList());

            long myOnes =default;
            List<char> myLastOxygen   ;        //seq[char]
            List<char> myLastCO2;                //seq[char]
            List<char> myReadings;              //seq [seq[char] ]
           string myBit;                //Char

            myReadings = new List<List<char>>(s.Data);
            

            
            for ( int myIndex = 1; myIndex <s.Transposed.Count; myIndex++ )
                
                myLastOxygen = myReadings.Last
                myOnes = s.Transposed.Item(myIndex).CountIt(cmpEQ.Deb("1"))
                
                if ( myOnes >= myReadings.Count - myOnes )
                    {myBit = '1';}
                else
                    {myBit = '0'};
                    
                

                myReadings = myReadings.FilterIt(x=>myBit, myIndex))
            
                if ( myReadings.Count = 1 Then
                    
                    myLastOxygen = myReadings.First
                    Exit for (
                    
                } (
                
                s.Transposed = myReadings.Transpose
                
            }


            myReadings = s.Data.Clone
            s.Transposed = s.Data.Transpose

            
            for ( myIndex = 1 To s.Transposed.LastIndex
                myLastCO2 = myReadings.Last
                myOnes = s.Transposed.Item(myIndex).Countit(cmpEQ.Deb("1"))
                if ( myOnes >= myReadings.Count - myOnes Then
                    
                    myBit = "0"
                    
                else:
                
                    myBit = "1"
                    
                    
                } (
                
                myReadings = myReadings.FilterIt(cmpEQ.Deb(myBit, myIndex))
            
                if ( myReadings.Count = 1 Then
                    myLastCO2 = myReadings.First
                    Exit for (
                    
                } (
                
                s.Transposed = myReadings.Transpose
                
            }

            myResult As Long
            myResult =Convert.ToInt32(myLastOxygen.ToString(vbNullString)) * Convert.ToInt32(myLastCO2.ToString(vbNullString));
            Console.WriteLine( "The answer to Day 03 Part 2 is 4636702. Found {0}", myResult)
        }
    }
}